initial structure for MongoDB Atlas recevier

zenmoto opened the pull request:
**Description:**
This is the initial scaffolding without actual function for a new receiver to pull metrics and logs from [MongoDB Atlas](https://www.mongodb.com/cloud/atlas)

This PR only creates an empty receiver and stubs, actually fetching metrics will come next, followed by pulling logs.

zenmoto committed:
initial structure for MongoDB Atlas recevier

zenmoto committed:
lint fix, add README

zenmoto committed:
disable linting of params on stub function

jpkrohling commented on cmd/configschema/go.mod:
```diff
@@ -626,6 +630,8 @@ replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/tcplo
 
 replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/udplogreceiver => ../../receiver/udplogreceiver
 
+replace github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver => ../../receiver/mongodbatlasreceiver
```
I feel bad for being extremely picky, but can you place this after `memcachedreceiver`?

jpkrohling commented on receiver/mongodbatlasreceiver/README.md:
```diff
@@ -0,0 +1,23 @@
+
+## Getting Started
+
+- `public_key`
```
Is there a recommendation about how those values should be stored?

jpkrohling commented on receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:
```diff
@@ -0,0 +1,104 @@
+		page++
+		if err != nil {
+			// Error was reported in getOrganizationsPage
+			break // Stop, returning what we have (probably empty slice)
```
Given that you are swallowing the error here, move the logging statement you have in the `getOrganizationsPage` to this place here. Make sure to use the `debug` level instead. Ideally, this failure would also be recorded as a gauge.

zenmoto replied:
I was just trying to toss in a metric here too, but I was having a difficult time finding an example of another receiver that does this. I gave myself a TODO for the moment while I look for good examples to emulate. I'll add a metric on the next PR with the client as I get it figured out. It should probably be a sum error count with the failed call as an attribute?

jpkrohling replied:
Just keep this as a todo somewhere, as I'm not sure we have sorted out observability for the collector yet (or have we, @bogdandrutu?)

jpkrohling commented on receiver/mongodbatlasreceiver/receiver.go:
```diff
@@ -0,0 +1,93 @@
+	ctx := s.ctx
+	for _, org := range s.client.Organizations(ctx) {
+		// Metrics collection starts here
+		s.log.Info("fetch resources for MongoDB Organization", zap.String("org", org.Name))
```
Debug instead?

rmfitzpatrick commented on receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:
```diff
@@ -0,0 +1,104 @@
+	t := digest.NewTransport(publicKey, privateKey)
+	tc, err := t.Client()
+	if err != nil {
+		return nil, err
```
Wrapping with some context can be helpful where logging isn't performed.  There are currently a couple similar scenarios where raw errors are returned.

```suggestion
		return nil, fmt.Errorf("could not create Mongo transport http client: %w", err)
```

zenmoto replied:
Thanks for this. I think I have all the current errors wrapped in the right places now.

zenmoto committed:
address pull request feedback

Added context to returned errors, reordered 'replace' entries in go.mod,
configuration detail in README.md

zenmoto committed:
added context to another error

jpkrohling commented on receiver/mongodbatlasreceiver/factory.go:
```diff
@@ -0,0 +1,59 @@
+	cfg := rConf.(*Config)
+	ms, err := newMongoDBAtlasReceiver(ctx, params.Logger, cfg, consumer)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create a MongoDB Atlas Receiver instance: %w", err)
```
```suggestion
		return nil, errors.Wrap(err, "unable to create a MongoDB Atlas Receiver instance")
```

jpkrohling commented on receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:
```diff
@@ -0,0 +1,105 @@
+	t := digest.NewTransport(publicKey, privateKey)
+	tc, err := t.Client()
+	if err != nil {
+		return nil, fmt.Errorf("Could not create MongoDB Atlas transport HTTP client: %w", err)
```
```suggestion
		return nil, errors.Wrap(err, "Could not create MongoDB Atlas transport HTTP client")
```

jpkrohling commented on receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go:
```diff
@@ -0,0 +1,105 @@
+	})
+	err = checkMongoDBClientErr(err, response)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error in retrieving organizations: %w", err)
```
```suggestion
		return nil, false, errors.Wrap(err, "error in retrieving organizations")
```

jpkrohling reviewed (approved):
LGTM, just a couple of minor things that can be improved in a follow-up PR.

zenmoto committed:
change from fmt.Error to errors.Wrap
